### PR TITLE
Barracuda inference for hybrid actions

### PR DESCRIPTION
**Pull request by dongruoping**
### Proposed change(s)

Barracuda inference for hybrid actions.

Major changes:

**1. The input argument `lastActions` to TensorApplier.IApplier is now an `ActionBuffers` instead of float[].**

**2. Changes to model output nodes**

To support hybrid actions, deprecated old action output nodes which only work for single-type actions and added new output nodes:
 - Deprecated `action` in favor of `continuous_actions` and `discrete_actions`
 - Similarly, deprecated `action_output_shape` in favor of `continuous_action_output_shape` and `discrete_action_output_shape`
 - `is_continuous_control` is also deprecated

Now the expectation of the model outputs is either set of:
 - (old case) `action`, `action_output_shape`, `is_continuous_control`
 - (new case) `continuous_actions`, `discrete_actions`, `continuous_action_output_shape`, `discrete_action_output_shape`

Running inference will use the new nodes by default, and fall back to old ones if they doesn't exist.
Deprecated nodes are still supported and the code is able to run old models.

Version compatibility:
old model, old C# - current behavior, use old output nodes
new model, old C# - use old output nodes just like current behavior, new nodes are ignored
old model, new C# - detect that new output nodes do not exist in the model and fall back to old ones
new model, new C# - use new output nodes

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs [hunk 1]
```diff
@@ -13,23 +14,32 @@ namespace Unity.MLAgents.Inference
...
                 if (lastActions.ContainsKey(agentId))
                 {
-                    var actionValue = lastActions[agentId];
-                    if (actionValue == null)
+                    var actionBuffer = lastActions[agentId];
+                    if (actionBuffer.Equals(ActionBuffers.Empty))
                     {
-                        actionValue = new float[actionSize];
-                        lastActions[agentId] = actionValue;
+                        actionBuffer = new ActionBuffers(new float[m_ActionSpec.NumContinuousActions],
```
Comment (dongruoping) [inline com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs:35]: What I don't like about this is ContinuousActionOutputApplier has to initialize new ActionBuffers based on discrete actions branchSizes.

Comment (chriselion) [reply com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs:35]: Maybe make another ActionBuffers constructor (or static method) that takes an ActionSpec?

File: com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs [hunk 2]
```diff
@@ -13,23 +14,32 @@ namespace Unity.MLAgents.Inference
...
             var actionSize = tensorProxy.shape[tensorProxy.shape.Length - 1];
             var agentIndex = 0;
             foreach (int agentId in actionIds)
             {
                 if (lastActions.ContainsKey(agentId))
                 {
-                    var actionValue = lastActions[agentId];
-                    if (actionValue == null)
+                    var actionBuffer = lastActions[agentId];
+                    if (actionBuffer.Equals(ActionBuffers.Empty))
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs:33]: Add an IsEmpty() method instead of comparing?

File: com.unity.ml-agents/Runtime/Policies/BarracudaPolicy.cs [hunk 3]
```diff
@@ -64,14 +63,8 @@ public void RequestDecision(AgentInfo info, List<ISensor> sensors)
...
             m_ModelRunner?.DecideBatch();
-            var actions = m_ModelRunner?.GetAction(m_AgentId);
-            if (m_SpaceType == SpaceType.Continuous)
-            {
-                m_LastActionBuffer = new ActionBuffers(actions, Array.Empty<int>());
-                return ref m_LastActionBuffer;
-            }
-
-            m_LastActionBuffer = ActionBuffers.FromDiscreteActions(actions);
+            var actionbuffer = m_ModelRunner?.GetAction(m_AgentId);
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Policies/BarracudaPolicy.cs:66]: Not sure how `?.` works with structs. Maybe best to write this as something like
```
if(m_ModelRunner == null) {
  return ActionBuffers.Empty;
}
m_ModelRunner.DecideBatch();
return m_ModelRunner.GetAction(m_AgentId);
```

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs [hunk 4]
```diff
@@ -627,13 +661,27 @@ static IEnumerable<string> CheckOutputTensorPresence(Model model, int memory)
             var tensorTester = new Dictionary<string, Func<BrainParameters, ActuatorComponent[], TensorShape?, int, int, string>>();
 
             // This will need to change a bit for hybrid action spaces.
-            if (isContinuous == ModelActionType.Continuous)
+            if (!model.outputs.Contains(TensorNames.ContinuousActionOutput) && !model.outputs.Contains(TensorNames.DiscreteActionOutput))
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:664]: Remove the above comment :)

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs [hunk 5]
```diff
@@ -338,7 +363,7 @@ SensorComponent[] sensorComponents
             }
 
             // If the model uses discrete control but does not have an input for action masks
-            if (isContinuous == ModelActionType.Discrete)
+            if (modelActionType == ModelActionType.Discrete || modelActionType == ModelActionType.Hybrid)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:366]: I think some of this code would be cleaner if ModelActionType were a Flags instead of Enum.

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs [hunk 6]
```diff
@@ -627,13 +661,27 @@ static IEnumerable<string> CheckOutputTensorPresence(Model model, int memory)
             var tensorTester = new Dictionary<string, Func<BrainParameters, ActuatorComponent[], TensorShape?, int, int, string>>();
 
             // This will need to change a bit for hybrid action spaces.
-            if (isContinuous == ModelActionType.Continuous)
+            if (!model.outputs.Contains(TensorNames.ContinuousActionOutput) && !model.outputs.Contains(TensorNames.DiscreteActionOutput))
             {
-                tensorTester[TensorNames.ActionOutput] = CheckContinuousActionOutputShape;
+                if (modelActionType == ModelActionType.Continuous)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:666]: I think a lot of the changes in BarracudaModelParamLoader.cs could be simplified with an some extension methods on `Model`; something like
```
bool HasContinuousOutputs()
int ContinuousOutputSize()
string ContinuousOutputName()
```
and similar for discrete. All of those should handle the "legacy" case transparently. Then this code would simplify to
```
if(model.HasContinuousOutputs) {
  tensorTester[model.ContinuousOutputName] = CheckContinuousActionOutputShape;
}
if(model.HadDiscreteOutputs) {
  tensorTester[model.DiscreteOutputName] = CheckDiscreteActionOutputShape;
}
```
It could also simplify some of the new logic at the start of CheckModel()

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs [hunk 7]
```diff
@@ -599,24 +635,22 @@ static IEnumerable<string> CheckOutputTensorPresence(Model model, int memory)
...
         {
             var failedModelChecks = new List<string>();
-            if (isContinuous == ModelActionType.Unknown)
-            {
-                failedModelChecks.Add("Cannot infer type of Control from the provided model.");
-                return failedModelChecks;
-            }
-            if (isContinuous == ModelActionType.Continuous &&
+
+            // TODO: Update this check after intergrating ActionSpec into BrainParameters
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:643]: If anything, we're going to be deprecating SpaceType, not adding Hybrid to it. 

I'm not sure how to keep these checks meaningful, since we can have a hybrid action but the Agent's BrainParameters can still be all continuous or all discrete.

File: com.unity.ml-agents/Runtime/Inference/TensorNames.cs [hunk 8]
```diff
@@ -23,12 +23,16 @@ internal static class TensorNames
...
+        public const string ContinuousActionOutputShape = "continuous_action_output_shape";
+        public const string DiscreteActionOutputShape = "discrete_action_output_shape";
+        public const string ActionOutputDeprecated = "action";
+        public const string ContinuousActionOutput = "continuous_actions";
+        public const string DiscreteActionOutput = "discrete_actions";
 
         public static readonly string[] RequiredConstants =
         {
-            VersionNumber, MemorySize, IsContinuousControl, ActionOutputShape
+            VersionNumber, MemorySize
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/TensorNames.cs:35]: These were added because if people passed a non-ML-Agents model without the required tensors, they would get obscure null reference errors when we tried to pull out the constants. For example, in 
```
var modelApiVersion = (int)model.GetTensorByName(TensorNames.VersionNumber)[0];
```
Now that now all of them are required, but the logic is more complicated (e.g. need ActionOutputShape OR ContinuousActionOutputShape OR DiscreteActionOutputShape), we should probably remove that block that checks this, and just make sure that the tensor is there before you use it (and raise a friendly message if it's not).

File: com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs (hunk 1 above)
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs:35]: Yes I can do that.
But even with that I'm still not able to initialize continuous/discrete field separately and still need both sizes to initialize the new ActionBuffers here, due to ActionBuffers's read-only properties. I wonder if it's reasonable to change that.

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs (hunk 7 above)
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:643]: In that case, maybe checking all the continuous/discrete sizes and branches to match between BrainParameters and the model is sufficient?
and just delete these checks

File: com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs (hunk 1 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs:35]: I don't think initializing them separately is a requirement; you have all the information you need in the ActionSpec. Encapsulating it in an ActionBuffers static method seems fine to me.

You could also move the logic to a ModelRunner method, and call it in ModelRunner.DecideBatch(), just before the call to m_TensorApplier.ApplyTensors.

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs (hunk 7 above)
Comment (chriselion) [reply com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:643]: I guess you can check that if `brainParameters.VectorActionSpaceType == SpaceType.Continuous` that the number of actions is <= the total number of model continuous actions (and similar for discrete)

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs [hunk 9]
```diff
@@ -182,19 +193,32 @@ public static string[] GetOutputNames(Model model)
...
+            ModelActionType modelActionType;
+            if (model.outputs.Contains(TensorNames.ContinuousActionOutputShape) || model.outputs.Contains(TensorNames.DiscreteActionOutputShape))
+            {
+                modelContinuousActionSize = (int)model.GetTensorByName(TensorNames.ContinuousActionOutputShape)[0];
+                modelDiscreteActionSize = (int)model.GetTensorByName(TensorNames.DiscreteActionOutputShape)[0];
+                modelActionType = GetActionType(modelContinuousActionSize > 0, modelDiscreteActionSize > 0);
+            }
+            else
+            {
+                var isContinuous = model.GetTensorByName(TensorNames.IsContinuousControl)[0] > 0;
```
[the remark this answers is not in the document]
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs]: yes, changed it as suggested

File: com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs (hunk 2 above)
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Inference/ApplierImpl.cs]: done

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs [hunk 10]
```diff
@@ -164,9 +178,6 @@ public static string[] GetOutputNames(Model model)
 
             var modelApiVersion = (int)model.GetTensorByName(TensorNames.VersionNumber)[0];
             var memorySize = (int)model.GetTensorByName(TensorNames.MemorySize)[0];
-            var isContinuousInt = (int)model.GetTensorByName(TensorNames.IsContinuousControl)[0];
```
[the remark this answers is not in the document]
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:167]: updated in the description

File: com.unity.ml-agents/Runtime/Inference/TensorNames.cs [hunk 11]
```diff
@@ -23,12 +23,16 @@ internal static class TensorNames
         public const string MemorySize = "memory_size";
         public const string VersionNumber = "version_number";
         public const string IsContinuousControl = "is_continuous_control";
-        public const string ActionOutputShape = "action_output_shape";
-        public const string ActionOutput = "action";
+        public const string ActionOutputShapeDeprecated = "action_output_shape";
```
[the remark this answers is not in the document]
Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Inference/TensorNames.cs:25]: Move ActionOutputShapeDeprecated at the bottom with ActionOutputDeprecated

File: com.unity.ml-agents/Tests/Editor/ParameterLoaderTest.cs [hunk 12]
```diff
@@ -157,7 +157,7 @@ public void TestGetOutputTensors1()
         {
             var model = ModelLoader.Load(continuous2vis8vec2actionModel);
```
Comment (vincentpierre) [inline com.unity.ml-agents/Tests/Editor/ParameterLoaderTest.cs:158]: I think you made a lot of improvements to the model loading logic. I think you should add more tests to check your logic (loading an old model and a new model and see if they behave properly)

**Review by vincentpierre — APPROVED** [review]

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        /// <summary>
+        /// Discrete action output size of the model.
+        /// </summary>
+        /// <param name="model">
+        /// The Barracuda engine model for loading static parameters.
+        /// </param>
+        /// <returns>Size of discrete action output.</returns>
+        public static int DiscreteOutputSize(this Model model)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/BarracudaModelExtensions.cs:96]: Might want to make it clearer (either in comments or method name) that this is the sum of the branch sizes, not the number of branches. That was confusing when its return value was assigned to `modelSumDiscreteBranchSizes` later.

File: com.unity.ml-agents/Runtime/Inference/TensorApplier.cs [hunk 14]
```diff
@@ -52,16 +52,25 @@ public interface IApplier
...
+                useDeprecated = model.UseDeprecated();
+            }
+            if (useDeprecated)
+            {
+                actionSpec.CheckNotHybrid();
+            }
             if (actionSpec.NumContinuousActions > 0)
             {
-                m_Dict[TensorNames.ActionOutput] = new ContinuousActionOutputApplier();
+                var tensorName = useDeprecated ? TensorNames.ActionOutputDeprecated : TensorNames.ContinuousActionOutput;
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/TensorApplier.cs:67]: `model.ContinuousOutputName()`?

File: com.unity.ml-agents/Runtime/Inference/TensorApplier.cs [hunk 15]
```diff
@@ -52,16 +52,25 @@ public interface IApplier
             Dictionary<int, List<float>> memories,
             object barracudaModel = null)
         {
-            actionSpec.CheckNotHybrid();
-
+            bool useDeprecated = false;
+            if (barracudaModel != null)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/TensorApplier.cs:56]: Can we just early out if the model is null? Might make some of the code cleaner. (same for TensorGenerator below)

File: com.unity.ml-agents/Runtime/Inference/TensorApplier.cs (hunk 14 above)
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Inference/TensorApplier.cs:67]: Looks like it's possible that the model is `null` here

File: com.unity.ml-agents/Runtime/Inference/TensorApplier.cs (hunk 15 above)
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Inference/TensorApplier.cs]: done

File: com.unity.ml-agents/Tests/Editor/ParameterLoaderTest.cs (hunk 12 above)
Comment (dongruoping) [reply com.unity.ml-agents/Tests/Editor/ParameterLoaderTest.cs]: More tests with new models added

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs [hunk 16]
```diff
@@ -72,12 +65,9 @@ public static int GetNumVisualInputs(Model model)
 
             foreach (var input in model.inputs)
             {
-                if (input.shape.Length == 4)
+                if (input.name.StartsWith(TensorNames.VisualObservationPlaceholderPrefix))
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:68]: Doesn't have to be in this PR, but you might want to consider moving these to the extension methods too.

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs [hunk 17]
```diff
@@ -152,21 +149,13 @@ public static string[] GetOutputNames(Model model)
                 return failedModelChecks;
             }
 
-            foreach (var constantName in TensorNames.RequiredConstants)
+            var modelApiVersionTensor = model.GetTensorByName(TensorNames.VersionNumber);
+            if (modelApiVersionTensor == null)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:153]: Again, doesn't need to be in this PR, but I feel like we lost some of the checks that the tensors we need are in the Model. It's never an issue for model files that we create, but a frequent source of forum/github posts would be people trying to train their own models, then add them for inference and get null reference errors because the assumed the tensors were there without checking.

I need to look a little more closely, but I think things like HasContinuousOutputs() would throw if not of the expected tensors are there.

We could add something like

```
bool CheckExpectedTensors(List<string> errorMessagesOut) {}
```
to the extension methods, call it once here, and then assume everything is present after that.

We should also have a test that CheckModel() with a basically empty model doesn't throw and returns some error strings.

Feel free to just log a jira as a reminder to come back to this, since I think it's a small regression in usability.

**Review by chriselion — APPROVED** [review]

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs (hunk 17 above)
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Inference/BarracudaModelParamLoader.cs:153]: I think I took care of that by the order of the checks, but I agree we should check it explicitly for safety. 
Made a [JIRA ticket](https://jira.unity3d.com/browse/MLA-1565).

**Comment by dongruoping** [comment]
Some tests are failing because since this PR is not self-contained, and it needs some changes on python. Leave those to python PR will check when merge.
